libcloud/compute/drivers/onapp.py

```diff
@@ -0,0 +1,842 @@
+                                    "Content-type": "application/json"},
+                                method="PUT")
+
+    #
+    # Methods to be implemented
+    #
+    def list_sizes(self, location=None):
+        pass
+
+    def list_locations(self):
+        pass
+
+    def reboot_node(self, node):
+        pass
+
+    def list_volumes(self):
+        pass
+
+    def list_volume_snapshots(self, volume):
+        pass
+
+    def create_volume(self, size, name, location=None, snapshot=None):
+        pass
+
+    def create_volume_snapshot(self, volume, name):
+        pass
+
+    def attach_volume(self, node, volume, device=None):
+        pass
+
+    def detach_volume(self, volume):
+        pass
+
+    def destroy_volume(self, volume):
+        pass
+
+    def destroy_volume_snapshot(self, snapshot):
+        pass
+
+    def list_images(self, location=None):
+        pass
+
+    def create_image(self, node, name, description=None):
+        pass
+
+    def delete_image(self, node_image):
+        pass
+
+    def get_image(self, image_id):
+        pass
+
+    def copy_image(self, source_region, node_image, name, description=None):
+        pass
+
+    def list_key_pairs(self):
+        pass
+
+    def get_key_pair(self, name):
+        pass
+
+    def create_key_pair(self, name):
+        pass
+
+    def import_key_pair_from_string(self, name, key_material):
+        pass
+
+    def delete_key_pair(self, key_pair):
+        pass
+
+    #
+    # Helper methods
+    #
+
+    def _to_node(self, data):
+        identifier = data["identifier"]
+        label = data["label"]
+        ip_addresses = []
+        for ip in data["ip_addresses"]:
+            ip_addresses.append(OnAppIpAddress(**ip["ip_address"]))
+
+        extra = OnAppNodeDriver._get_extra_dict(
+            data, RESOURCE_EXTRA_ATTRIBUTES_MAP["node"]
+        )
+        return OnAppNode(identifier, label, ip_addresses, extra, self)
+
+    @staticmethod
+    def _get_extra_dict(response, mapping):
+        """
+        Extract attributes from the element based on rules provided in the
+        mapping dictionary.
+
+        :param   response: The JSON response to parse the values from.
+        :type    response: ``dict``
+
+        :param   mapping: Dictionary with the extra layout
+        :type    mapping: ``dict``
+
+        :rtype:  ``dict``
+        """
+        extra = {}
+        for attribute, values in mapping.items():
+            transform_func = values["transform_func"]
+            value = response.get(values["key_name"], None)
+
+            if value is not None:
+                extra[attribute] = transform_func(value)
+            else:
+                extra[attribute] = None
+        return extra
+
+    @staticmethod
+    def _create_args_to_params(**kwargs):
+        """
+        Extract server params from keyword args to create a VS
+
+        :param   kwargs: keyword args
+        :return: ``dict``
+        """
+        server_params = dict()
```

Comment:
Looks like the size of this method could be reduced drastically by using a for loop (we follow a similar pattern in some other drivers) :)